### PR TITLE
Implement Channel encryption

### LINKED ISSUE #18: Implement channel encryption
This requires the storage server to have access to the service node's key pair.

### PR DESCRIPTION
**Pull request by sachaaaaa**
Decent sized pull request, commits were neatly organised to help break down the changes.
closes #18

### CODE REVIEW

File: crypto/src/channel_encryption.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    ciphertext_len += len;
+
+    // Remove excess padding
+    output.resize(ciphertext_len);
+
+    // Insert iv at the start
+    output.insert(output.begin(), iv, iv + ivLength);
+
+    EVP_CIPHER_CTX_free(ctx);
+    //    const T ret(output.begin(), output.end());
```
Comment (BeaudanBrown) [inline crypto/src/channel_encryption.cpp]: This not needed now?

File: crypto/src/lokinet_identity.hpp [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+std::vector<uint8_t> parseLokinetIdentityPrivate(const std::string& path);
```
[the remark this answers is not in the document]
Comment (sachaaaaa) [reply crypto/src/lokinet_identity.hpp]: Thanks github to highlight it :)
fixed!

File: crypto/src/channel_encryption.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    if (crypto_sign_ed25519_sk_to_curve25519(this->privateKey.data(),
+                                             privateEd25519Key.data()) != 0) {
+        throw std::runtime_error(
+            "Could not convert lokinet private key from ed25519 to curve25519");
+    }
+}
+
+template <typename T>
+std::vector<uint8_t> ChannelEncryption<T>::calculateSharedSecret(
+    const std::vector<uint8_t>& pubkey) const {
```
[the remark this answers is not in the document]
Comment (sachaaaaa) [reply crypto/src/channel_encryption.cpp]: Good catch, one victory for consistency.

File: crypto/src/channel_encryption.cpp (hunk 1 above)
Comment (sachaaaaa) [reply crypto/src/channel_encryption.cpp]: Oops.
Removed!

File: httpserver/http_connection.hpp [hunk 4]
```diff
@@ -120,21 +125,18 @@ class http_connection : public std::enable_shared_from_this<http_connection> {
...
-
-        for (auto seq : request_.body().data()) {
-            const auto* cbuf = boost::asio::buffer_cast<const char*>(seq);
-            bytes.insert(std::end(bytes), cbuf,
-                         cbuf + boost::asio::buffer_size(seq));
+        try {
+            const std::string decoded = boost::beast::detail::base64_decode(plainText);
+            plainText = channelCipher_.decrypt(decoded, header_[LOKI_EPHEMKEY_HEADER]);
+        } catch(...) {
+            // TODO: don't accept unencrypted data?
```
[the remark this answers is not in the document]
Comment (sachaaaaa) [reply httpserver/http_connection.hpp]: Good point. Added error handling.

File: httpserver/main.cpp [hunk 5]
```diff
@@ -20,19 +21,20 @@ int main(int argc, char* argv[]) {
...
+        std::cout << "Listening at address " << argv[1] << " port " << argv[2]
+                  << std::endl;
 
         auto const address = boost::asio::ip::make_address(argv[1]);
         unsigned short port = static_cast<unsigned short>(std::atoi(argv[2]));
 
         boost::asio::io_context ioc{1};
 
         Storage storage(".");
+        ChannelEncryption<std::string> channelEncryption("/Users/sachav/Downloads/identity.private");
```
[the remark this answers is not in the document]
Comment (sachaaaaa) [reply httpserver/main.cpp]: Done in latest commit!

**Pull request merged**